e2e/specs/websocket_reconnects.spec.js

```diff
@@ -0,0 +1,36 @@
+/**
+ * Copyright (c) Streamlit Inc. (2018-2022) Snowflake Inc. (2022)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+describe("websocket reconnects", () => {
+  beforeEach(() => {
+    cy.loadApp("http://localhost:3000/");
+  });
+
+  it("persists session state when the websocket connection is dropped and reconnects", () => {
+    for (let i = 0; i < 5; i++) {
+      cy.get(".stButton button").contains("click me!").click();
+    }
+
+    cy.window().then((win) => {
+      win.streamlitDebug.disconnectWebsocket();
+
+      // Wait until we've reconnected and rerun the script.
+      cy.get("[data-testid='stStatusWidget']").should("not.exist");
```

Comment:
This is a bit tricky because all of the easily observable things that tell us that we've disconnected and reconnected happen for a very brief time that could easily be missed when watching the DOM (and I don't think there's another easy/sane way to get visibility into the connection state)

I did verify that this test consistently fails in the previous, pre-reconnect improvements world, but I think this can be made more robust by doing something like

1) setting a timeout to disconnect the websocket in, say, 100ms
2) waiting to see the `stStatusWidget` appear to confirm the disconnect
3) then waiting for it to disappear again to confirm the reconnect

I'm worried that missing step 2 will result in this test flaking out frequently, but let me go ahead and change the test to do this, and we can consider changing it back if flakiness does increase a ton